`expected` compilation errors

Keep seeing this anytime I try to get the result of `Extract` into a `unique_ptr` object using `std::move`.

```
expected.hpp:666:39: error: could not convert 
‘((opentracing::expected<std::unique_ptr<opentracing::v1::SpanContext> >*)this)- >
opentracing::expected<std::unique_ptr<opentracing::v1::SpanContext>>::contained.opentracing::
expected_detail::storage_t<std::unique_ptr<opentracing::v1::SpanContext>, std::error_code>::error()’ 
from ‘opentracing::expected_detail::storage_t<std::unique_ptr<opentracing::v1::SpanContext>, 
std::error_code>::error_type {aka std::error_code}’ to ‘std::__exception_ptr::exception_ptr’
```

--- a/3rd_party/include/opentracing/expected/expected.hpp
+++ b/3rd_party/include/opentracing/expected/expected.hpp
@@ -678,16 +678,6 @@ public:
         return has_value_;
     }
 
-    nsel_REQUIRES_0(!std::is_same<E, std::exception_ptr>::value)
-
-    constexpr value_type const & value() const &
-    {
-      return has_value()
-                 ? contained.value()
-                 : (throw bad_expected_access<error_type>(contained.error()),
-                    contained.value());
-    }
-
     constexpr value_type const & value() const &
     {
         return has_value()

--- a/3rd_party/include/opentracing/expected/expected.hpp
+++ b/3rd_party/include/opentracing/expected/expected.hpp
@@ -41,6 +41,9 @@
 #define nsel_REQUIRES_0(...) \
     template< bool B = (__VA_ARGS__), typename std::enable_if<B, int>::type = 0 >
 
+#define nsel_REQUIRES_1(...) \
+  template <int B = (__VA_ARGS__), typename std::enable_if<B, int>::type = 0>
+
 #define nsel_REQUIRES_T(...) \
     typename = typename std::enable_if< (__VA_ARGS__), opentracing::expected_detail::enabler >::type
 
@@ -188,7 +191,7 @@ private:
 
 /// class unexpected_type
 
-template< typename E = std::exception_ptr >
+template< typename E = std::error_code >
 class unexpected_type
 {
 public:
@@ -649,31 +652,61 @@ public:
         return has_value_;
     }
 
+    nsel_REQUIRES_0(!std::is_same<E, std::exception_ptr>::value)
+
     constexpr value_type const & value() const &
     {
-        return has_value()
-            ? contained.value()
-            : std::is_same<error_type, std::exception_ptr>::value
-            ? ( std::rethrow_exception( contained.error() ), contained.value() )
-            : ( throw bad_expected_access<error_type>( contained.error() ), contained.value() );
+      return has_value()
+                 ? contained.value()
+                 : (throw bad_expected_access<error_type>(contained.error()),
+                    contained.value());
     }
+
+    nsel_REQUIRES_1(std::is_same<E, std::exception_ptr>::value)
+
+    constexpr value_type const & value() const &
+    {
+      return has_value() ? contained.value()
+                         : (std::rethrow_exception(contained.error()),
+                            contained.value());
+    }
+
+    nsel_REQUIRES_0(!std::is_same<E, std::exception_ptr>::value)
 
     value_type & value() &
     {
-        return has_value()
-            ? contained.value()
-            : std::is_same<error_type, std::exception_ptr>::value
-            ? ( std::rethrow_exception( contained.error() ), contained.value() )
-            : ( throw bad_expected_access<error_type>( contained.error() ), contained.value() );
+      return has_value()
+                 ? contained.value()
+                 : (throw bad_expected_access<error_type>(contained.error()),
+                    contained.value());
     }
 
-    constexpr value_type && value() const &&
+    nsel_REQUIRES_1(std::is_same<E, std::exception_ptr>::value)
+
+    value_type & value() &
     {
-        return has_value()
-            ? std::move( contained.value() )
-            : std::is_same<error_type, std::exception_ptr>::value
-            ? ( std::rethrow_exception( contained.error() ), contained.value() )
-            : ( throw bad_expected_access<error_type>( contained.error() ), contained.value() );
+      return has_value() ? contained.value()
+                         : (std::rethrow_exception(contained.error()),
+                            contained.value());
+    }
+
+    nsel_REQUIRES_0(!std::is_same<E, std::exception_ptr>::value)
+
+    constexpr value_type && value() &&
+    {
+      return has_value()
+                 ? std::move(contained.value())
+                 : (throw bad_expected_access<error_type>(contained.error()),
+                    contained.value());
+    }
+
+    nsel_REQUIRES_1(std::is_same<E, std::exception_ptr>::value)
+
+    constexpr value_type && value() &&
+    {
+      return has_value() ? std::move(contained.value())
+                         : (std::rethrow_exception(contained.error()),
+                            contained.value());
     }
 
     constexpr error_type const & error() const &
@@ -1208,6 +1241,7 @@ struct hash< opentracing::expected<void,E> >
 
 #undef nsel_REQUIRES
 #undef nsel_REQUIRES_0
+#undef nsel_REQUIRES_1
 #undef nsel_REQUIRES_T
 
 #endif // OPENTRACING_EXPECTED_LITE_HPP

--- a/3rd_party/include/opentracing/expected/expected.hpp
+++ b/3rd_party/include/opentracing/expected/expected.hpp
@@ -41,9 +41,6 @@
 #define nsel_REQUIRES_0(...) \
     template< bool B = (__VA_ARGS__), typename std::enable_if<B, int>::type = 0 >
 
-#define nsel_REQUIRES_1(...) \
-  template <int B = (__VA_ARGS__), typename std::enable_if<B, int>::type = 0>
-
 #define nsel_REQUIRES_T(...) \
     typename = typename std::enable_if< (__VA_ARGS__), opentracing::expected_detail::enabler >::type
 
@@ -387,6 +384,35 @@ private:
     error_type m_error;
 };
 
+/// class error_traits
+
+template< typename Error >
+struct error_traits
+{
+    static void rethrow( Error const & e )
+    {
+        throw bad_expected_access<Error>{ e };
+    }
+};
+
+template<>
+struct error_traits< std::exception_ptr >
+{
+    static void rethrow( std::exception_ptr const & e )
+    {
+        std::rethrow_exception( e );
+    }
+};
+
+template<>
+struct error_traits< std::error_code >
+{
+    static void rethrow( std::error_code const & e )
+    {
+        throw std::system_error( e );
+    }
+};
+
 /// class expected
 
 template< typename T, typename E = std::error_code >
@@ -662,51 +688,25 @@ public:
                     contained.value());
     }
 
-    nsel_REQUIRES_1(std::is_same<E, std::exception_ptr>::value)
-
     constexpr value_type const & value() const &
     {
-      return has_value() ? contained.value()
-                         : (std::rethrow_exception(contained.error()),
-                            contained.value());
+        return has_value()
+            ? ( contained.value() )
+            : ( error_traits<error_type>::rethrow( contained.error() ), contained.value() );
     }
-
-    nsel_REQUIRES_0(!std::is_same<E, std::exception_ptr>::value)
 
     value_type & value() &
     {
-      return has_value()
-                 ? contained.value()
-                 : (throw bad_expected_access<error_type>(contained.error()),
-                    contained.value());
+        return has_value()
+            ? ( contained.value() )
+            : ( error_traits<error_type>::rethrow( contained.error() ), contained.value() );
     }
-
-    nsel_REQUIRES_1(std::is_same<E, std::exception_ptr>::value)
-
-    value_type & value() &
-    {
-      return has_value() ? contained.value()
-                         : (std::rethrow_exception(contained.error()),
-                            contained.value());
-    }
-
-    nsel_REQUIRES_0(!std::is_same<E, std::exception_ptr>::value)
 
     constexpr value_type && value() &&
     {
-      return has_value()
-                 ? std::move(contained.value())
-                 : (throw bad_expected_access<error_type>(contained.error()),
-                    contained.value());
-    }
-
-    nsel_REQUIRES_1(std::is_same<E, std::exception_ptr>::value)
-
-    constexpr value_type && value() &&
-    {
-      return has_value() ? std::move(contained.value())
-                         : (std::rethrow_exception(contained.error()),
-                            contained.value());
+        return has_value()
+            ? ( contained.value() )
+            : ( error_traits<error_type>::rethrow( contained.error() ), contained.value() );
     }
 
     constexpr error_type const & error() const &
@@ -1241,7 +1241,6 @@ struct hash< opentracing::expected<void,E> >
 
 #undef nsel_REQUIRES
 #undef nsel_REQUIRES_0
-#undef nsel_REQUIRES_1
 #undef nsel_REQUIRES_T
 
 #endif // OPENTRACING_EXPECTED_LITE_HPP

